### ISSUE TITLE
add support for linux

### ISSUE THREAD

**Issue by Pevernow** [pr0g/sdl-bgfx-imgui-starter#1]
It does not work in ubuntu 20.04

**Comment by pr0g** [comment]
Hi @Pevernow,

Thanks for letting me know, it was always a long shot if it would as I've only tested the starter on macOS and Windows. 

Do you have any more information about what step failed? Did one of the third party libraries fail to build, was there a problem with CMake or did the main example not build or run?

You might need to build the shaders for OpenGL on Linux as bgfx uses Metal on macOS and DirectX on Windows. You can probably create a new `compile_shaders.sh` to do this.

Basically any more info would help 🙂 

Thanks!


Tom

**Comment by Pevernow** [comment]
I compiled it with cmake.
Shaders are also compiled.
But there was an error at run time.

Because there is no native code to support Linux specific environment，
I added this line:  pd.nwh = &wmi.info.x11.window;

There was a problem running to this line:
bgfx::init(bgfx_init);

X Error of failed request:  GLXBadDrawable
  Major opcode of failed request:  155 (GLX)
  Minor opcode of failed request:  5 (X_GLXMakeCurrent)
  Serial number of failed request:  29
  Current serial number in output stream:  29

The OS can running Minecraft successful.(Using Opengl)
So I think there is nothing wrong with my computer or devices

**Comment by pr0g** [comment]
Hi @Pevernow 

Thank you very much for following up on this! This info is great, super helpful to know it's a runtime thing.

I'll see if I can get a Linux instance running via a virtual machine to see if I can get a repo and report back. It sounds like OpenGL is failing to be found potentially, not sure if it's a hardware thing specifically or the OS. I'm sure we can track it down. 

If you check in the ReadMe.md there's two other blog posts which have info about setting up bgfx on Linux which might have some info too.

I'll let you know what I find,

Cheers!


Tom

**Comment by pr0g** [comment]
Making some progress...

Tried using VirtualBox to get Ubuntu working on my Mac but blasted macOS Big Sur put a stop to that 🙈 Fortunately, I have a copy of Parallels for using Windows and that also supports Ubuntu which is handy!

I had to install `cmake` and `git` first and then when trying to configure/build the third-party dependencies I got a failure in the `git-clone.sh` file in the `cmakefied` repo (for getting `Dear ImGui`). After some digging, I realized the reason for this was I'm using `sh`, not `bash` to invoke the script in the `CMakeLists.txt` file in the third-party folder (I'll push this fix shortly).

I also had to then install `X11` and `OpenGL` on the system as `bgfx` complains if it can't find them when configuring. Things are building now, `bgfx` takes a little while... Hopefully, I'll be able to get to your repro soonish...

**Comment by pr0g** [comment]
Hey @Pevernow 

Linux support has been added :slightly_smiling_face: 

See fa6534b4a5b96104e9568f80b5a7dbc302b6c6ba and 93f138b98af8150e3e1bc3a7691b3fe0764873a4 for more details.

Can you give things another try and let me know how you get on? I've been able to get things working locally.

![image](https://user-images.githubusercontent.com/1136625/103444205-517e6d00-4c1b-11eb-9e9c-ca9a1c79c59f.png)

You need to use `compile-shaders-linux.sh` now but everything else should be the same.

Once you've confirmed it's working I'll close the issue.

Cheers!

Tom

**Comment by Pevernow** [comment]
Thanks for your hard work.

No it works well.

**Comment by pr0g** [comment]
Awesome stuff! 😄 Thanks for letting me know 👍

No worries at all, have a good one!

Tom
